### PR TITLE
TF-TRT improve documentation of execution context handling

### PR DESCRIPTION
**Pull request by tfeher**
This PR improves the description of the execution context handling by TF-TRT.

The synchronization related to execution context management was discussed in issue #36959. Relevant parts from that discussion are added as comments to the EngineContext class.

Tagging @bixia1 for review.

### CODE REVIEW

File: tensorflow/compiler/tf2tensorrt/utils/trt_lru_cache.h [hunk 1]
```diff
@@ -146,10 +146,19 @@ struct EngineContext {
   }
 
   // In explicit batch mode, we maintain a vector of contexts for each engine,
-  // where each context is created for a different profile. The
+  // where each context is created for a different profile. The are two reasons
+  // for this:
+  // - In TRT 6 it is not possible to switch profile after it was set
```
Comment (bixia1) [inline tensorflow/compiler/tf2tensorrt/utils/trt_lru_cache.h:151]: s/profile after it was/profiles after they are/  or  s/profile after it was/a profile after it is/

File: tensorflow/compiler/tf2tensorrt/utils/trt_lru_cache.h [hunk 2]
```diff
@@ -146,10 +146,19 @@ struct EngineContext {
...
+  //   calls in that context are finished. This would require an additional
+  //   synchronization before we call setOptimizationProfile. To avoid this
+  //   extra sync call, we mantain separate execution context for each profile.
   // IExecutionContext object is not thread safe: only one thread should use it
   // for inference at a time therefore we need a mutex. More details at
   // https://docs.nvidia.com/deeplearning/sdk/tensorrt-best-practices/index.html#thread-safety
+  // Additional discussion about execution context management and thread safety
+  // at https://github.com/tensorflow/tensorflow/issues/36959
   std::vector<TrtUniquePtrType<nvinfer1::IExecutionContext>> execution_context
       TF_GUARDED_BY(mu);
```
Comment (bixia1) [inline tensorflow/compiler/tf2tensorrt/utils/trt_lru_cache.h:163]: I understand that we can't allow two threads to "use" the same execution context concurrently. What I don't understand here is that the mutex doesn't seem to protect each context, but rather protect accessing the vector of contexts. Is the following scenario possible?
Time-1: Thread-A retrieves Context-1 from the vector (through [here](https://github.com/tensorflow/tensorflow/blob/0a3ce8d533dc86ea2453017dbac915ab5d88e79e/tensorflow/compiler/tf2tensorrt/utils/trt_lru_cache.h#L137) )
Time-2: Thread-B retrieves Context-1 from the vector
Time-3: both Thread-A and Thread-B are using Context-1 for inference

Comment (tfeher) [reply tensorflow/compiler/tf2tensorrt/utils/trt_lru_cache.h:164]: The way we are using it is IExecutionContext is correct: we [hold the lock](https://github.com/tensorflow/tensorflow/blob/15e0f259b119f53b90fc1698f7abd9b739df3b6c/tensorflow/compiler/tf2tensorrt/kernels/trt_engine_op.cc#L757-L783) until we use the IExecutionContext object. After the inference jobs are enqued, the single CUDA compute stream ensures that there could be no concurrent inferences running in the same context (#36959). 

I do not know what is the best way to express this with thread safety annotations. The scenario that you write would be problematic. But looking at the definition of the [annotation macros](https://github.com/tensorflow/tensorflow/blob/15e0f259b119f53b90fc1698f7abd9b739df3b6c/tensorflow/core/platform/thread_annotations.h#L47-L64), I do not know what would be the correct way to express that the elements inside a container should be guarded.

File: tensorflow/compiler/tf2tensorrt/utils/trt_lru_cache.h (hunk 1 above)
Comment (tfeher) [reply tensorflow/compiler/tf2tensorrt/utils/trt_lru_cache.h]: Fixed.

**Review by tfeher — COMMENTED** [review]
Thanks @bixia1 for the comments, I have addressed the simple ones. Let me know if you have a suggestion on improving the thread safety annotations.

**Review by bixia1 — APPROVED** [review]

**Pull request merged**